text/0069-api-v2.md

```diff
@@ -0,0 +1,111 @@
+# TiKV API V2
+
+## Motivation
+
+`API V2` is a set of breaking changes that aim to solve serval issues with current RawKV (hereafter referred to as `API V1`):
+
+1. RawKV is not safe to use along with TxnKV.
+2. RawKV TTL is controlled by Store configuration. Switching the configuration will cause data corruption in silence.
+3. RawKV TTL is encoded into the value by appending 8-bytes UNIX timestamp to the end of the value, therefore it's hard to introduce other encode afterward.
+4. RawKV compare_and_swap is easy to make mistake.
+
+By solving issue 1, TiDB will be able to support RawKV as the table's storage engine, which will enrich TiDB's use cases.
+
+## Document-level design
+
+This RFC introduces a new key encode to RawKV and TxnKV, and a new value encode to RawKV, which will allow the RawKV to be used along with TxnKV and also allow TiKV to flexibly add meta, e.g. the TTL, to a RawKV value.
+
+The `API V2` is enabled by a switch on PD. Since it changes the storage encode, it will be not compatible to switch between `API V1` and `API V2` while there are non-TiDB data in TiKV. TiDB data is specially treated in order to not be affected by the change.
+
+### Upgrade
+
+Upgrade from `API V1` to `API V2` is a simple process:
+
+1. Update TiKV, TiDB, and PD to the version that supports `API V2`.
+2. Ensure that no non-TiDB data is in TiKV. Detele if any. (guaranteed by user)
+3. Ensure that no non-TiDB client is online. (guaranteed by user)
+4. Use `pd-ctl` to enable `API V2`.
+5. Restart all TiKV. (guaranteed by user)
+
+### Downgrade
+
+Downgrade from `API V2` to `API V1` is also simple:
+
+1. Ensure that no non-TiDB TiKV client is online.
+2. Ensure that no non-TiDB data is in TiKV. Detele if any. (guaranteed by user)
+3. Use `pd-ctl` to disable `API V2`.
+4. Restart all TiKV. (guaranteed by user)
+
+### Client
+
+TiKV clients will have a new configuration choosing `API V1` or `API V2` mode on connecting to TiKV.
+
+TiKV will reject the request if the client mode is mismatched to the TiKV configuration. There is one exception: With TiKV in `API V2` mode, you can use `TxnClient` in `API V1` mode to access the TiDB data (key starting with `m` or `t`) which is invisible to the client in `API V2` mode.
+
+Note that in `API V2` mode, you can't specify CF in `RawClient`.
+
+Using TxnKV along with RawKV is safe in `API V2` mode.
+
+## Reference-level design
```

Comment:
could you add a section to illustrate the compatibility issues, consider these scenarios:

- (client v1, TiKV API v2, TxnKV)
- (client v1, TiKV API v2, RawKV)
- (client v2, TiKV API v1, TxnKV)
- (client v2, TiKV API v1, RawKV)

the following information needs to be provided:

- whether the scenario is supported
- if not, is there a way to work around, and how to workaround?

Reply:
updated